File: scripts/seed.py
#!/usr/bin/env python3
"""Seed script for Vibe Kanban."""

import asyncio

import aiosqlite
import structlog

from vibe_kanban_clone.config import settings
from vibe_kanban_clone.db.base import Base
from vibe_kanban_clone.db.engine import engine
from vibe_kanban_clone.schemas.agent import AgentCreate
from vibe_kanban_clone.schemas.column import ColumnCreate
from vibe_kanban_clone.schemas.project import ProjectCreate
from vibe_kanban_clone.schemas.skill import SkillCreate
from vibe_kanban_clone.schemas.task import TaskCreate
from vibe_kanban_clone.schemas.task_comment import TaskCommentCreate
from vibe_kanban_clone.services import agents, columns, comments, projects, skills, tasks

logger = structlog.get_logger()


async def main() -> None:
    """Seed the database with sample data."""
    db_str = str(settings.db_path)
    async with aiosqlite.connect(db_str) as conn:
        await conn.execute("PRAGMA journal_mode=WAL")
        await conn.commit()

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.create_all)

    from sqlalchemy.ext.asyncio import async_sessionmaker

    AsyncSessionLocal = async_sessionmaker(engine, expire_on_commit=False)

    async with AsyncSessionLocal() as session:
        project = await projects.create_project(
            session, ProjectCreate(name="Vibe Kanban", description="Main board")
        )

        column_specs = [
            ("Backlog", 0, False),
            ("Todo", 1, False),
            ("In Progress", 2, False),
            ("Review", 3, False),
            ("Done", 4, True),
        ]
        created_columns = []
        for name, pos, terminal in column_specs:
            col = await columns.create_column(
                session,
                project.id,
                ColumnCreate(name=name, position=pos, is_terminal=terminal),
            )
            created_columns.append(col)

        agent_names = [
            "Alpha-Bot",
            "Beta-Bot",
            "Gamma-Bot",
            "Delta-Bot",
            "Epsilon-Bot",
        ]
        created_agents = []
        for name in agent_names:
            agent = await agents.create_agent(
                session,
                AgentCreate(name=name, system_prompt=f"You are {name}."),
            )
            created_agents.append(agent)

        skill_specs = [
            ("Research", "Find information", "Search and summarize", ["web_search"]),
            ("Code", "Write code", "Implement features", ["code_editor"]),
            ("Review", "Code review", "Review pull requests", ["code_editor", "git"]),
            ("Test", "Write tests", "Unit and integration tests", ["pytest"]),
            ("Deploy", "Deploy services", "CI/CD pipelines", ["docker", "kubernetes"]),
            ("Design", "UI/UX design", "Create mockups", ["figma"]),
            ("Docs", "Documentation", "Write docs and guides", ["markdown"]),
            ("Analytics", "Data analysis", "Analyze metrics", ["sql", "python"]),
        ]
        created_skills = []
        for name, desc, instr, tools in skill_specs:
            skill = await skills.create_skill(
                session,
                SkillCreate(name=name, description=desc, instructions=instr, allowed_tools=tools),
            )
            created_skills.append(skill)

        for i, agent in enumerate(created_agents):
            skill = created_skills[i % len(created_skills)]
            await agents.assign_skill(session, agent, skill)

        task_titles = [
            "Set up repo",
            "Design schema",
            "Implement models",
            "Write migrations",
            "Build API routes",
            "Add auth middleware",
            "Create React app",
            "Set up Tailwind",
            "Build Kanban board",
            "Add drag and drop",
            "Integrate WebSockets",
            "Write unit tests",
            "Write integration tests",
            "Set up CI",
            "Deploy to staging",
            "Performance audit",
            "Add logging",
            "Error handling",
            "Write documentation",
            "User feedback loop",
            "Finalize release",
            "Celebrate launch",
        ]
        created_tasks = []
        for idx, title in enumerate(task_titles):
            col = created_columns[idx % len(created_columns)]
            agent = created_agents[idx % len(created_agents)]
            task = await tasks.create_task(
                session,
                TaskCreate(
                    project_id=project.id,
                    column_id=col.id,
                    title=title,
                    description=f"Details for {title}",
                    priority=(idx % 4) + 1,
                    assigned_agent_id=agent.id,
                ),
            )
            created_tasks.append(task)

        comment_specs = [
            (created_tasks[0].id, "Alice", "Great start!"),
            (created_tasks[1].id, "Bob", "Needs more indexes."),
            (created_tasks[2].id, "Charlie", "Looks good to me."),
        ]
        for task_id, author, body in comment_specs:
            await comments.create_comment(
                session, task_id, TaskCommentCreate(author=author, body=body)
            )

    logger.info("database_seeded", tasks=len(task_titles), agents=len(agent_names))


if __name__ == "__main__":
    asyncio.run(main())

File: server/src/prisma/seed.ts
import { config } from "dotenv";
import path from "node:path";
config({ path: path.resolve(import.meta.dirname, "../../.env") });

import { prisma } from "../prisma/client.js";

async function main() {
  try {
    await prisma.$queryRawUnsafe("PRAGMA journal_mode=WAL;");
  } catch {
    // Ignore — may fail if driver disallows result-returning pragmas in executeRaw
  }

  // Clean existing data (order matters for FK constraints)
  const tables = [
    "task_runs",
    "task_comments",
    "agent_tools",
    "agent_skills",
    "tasks",
    "columns",
    "tools",
    "skills",
    "agents",
    "models",
    "projects",
  ];
  for (const t of tables) {
    await prisma.$executeRawUnsafe(`DELETE FROM ${t};`);
  }

  const project = await prisma.project.create({
    data: { name: "Agent Board", description: "Main board", slug: "agent-board" },
  });

  const columns = await prisma.$transaction([
    prisma.column.create({ data: { project_id: project.id, name: "Backlog", position: 0, is_terminal: false } }),
    prisma.column.create({ data: { project_id: project.id, name: "Todo", position: 1, is_terminal: false } }),
    prisma.column.create({ data: { project_id: project.id, name: "In Progress", position: 2, is_terminal: false } }),
    prisma.column.create({ data: { project_id: project.id, name: "Review", position: 3, is_terminal: false } }),
    prisma.column.create({ data: { project_id: project.id, name: "Done", position: 4, is_terminal: true } }),
  ]);

  const agentNames = [
    "Alpha-Bot",
    "Beta-Bot",
    "Gamma-Bot",
    "Delta-Bot",
    "Epsilon-Bot",
  ];
  const agents = [];
  for (const name of agentNames) {
    agents.push(
      await prisma.agent.create({
        data: { name, system_prompt: `You are ${name}.` },
      })
    );
  }

  const skillSpecs = [
    { name: "Research", description: "Find information", instructions: "Search and summarize", allowed_tools: JSON.stringify(["web_search"]) },
    { name: "Code", description: "Write code", instructions: "Implement features", allowed_tools: JSON.stringify(["code_editor"]) },
    { name: "Review", description: "Code review", instructions: "Review pull requests", allowed_tools: JSON.stringify(["code_editor", "git"]) },
    { name: "Test", description: "Write tests", instructions: "Unit and integration tests", allowed_tools: JSON.stringify(["pytest"]) },
    { name: "Deploy", description: "Deploy services", instructions: "CI/CD pipelines", allowed_tools: JSON.stringify(["docker", "kubernetes"]) },
    { name: "Design", description: "UI/UX design", instructions: "Create mockups", allowed_tools: JSON.stringify(["figma"]) },
    { name: "Docs", description: "Documentation", instructions: "Write docs and guides", allowed_tools: JSON.stringify(["markdown"]) },
    { name: "Analytics", description: "Data analysis", instructions: "Analyze metrics", allowed_tools: JSON.stringify(["sql", "python"]) },
  ];
  const skills = [];
  for (const s of skillSpecs) {
    skills.push(await prisma.skill.create({ data: s }));
  }

  const toolSpecs = [
    { name: "Read File", description: "Read the contents of a file or list directory entries.", handler_key: "read" },
    { name: "Edit File", description: "Replace an exact string in a file.", handler_key: "edit" },
    { name: "Write File", description: "Create or overwrite a file.", handler_key: "write" },
    { name: "Apply Patch", description: "Apply a patch to add, update, move, or delete files.", handler_key: "apply_patch" },
    { name: "Shell", description: "Execute a shell command and return its output.", handler_key: "shell" },
    { name: "Background Task", description: "Spawn a shell command as a background task.", handler_key: "task" },
    { name: "Task Status", description: "Check the status of a background task by ID.", handler_key: "task_status" },
    { name: "Web Search", description: "Search the web using DuckDuckGo.", handler_key: "websearch" },
    { name: "Web Fetch", description: "Fetch a URL and return its text content.", handler_key: "webfetch" },
    { name: "Repo Overview", description: "Summarize a repository structure.", handler_key: "repo_overview" },
    { name: "Repo Clone", description: "Clone a git repository.", handler_key: "repo_clone" },
    { name: "Glob Search", description: "Find files matching a glob pattern.", handler_key: "glob" },
    { name: "Grep Search", description: "Search file contents with regex.", handler_key: "grep" },
    { name: "Todo List", description: "Manage a todo list for the current run.", handler_key: "todo" },
    { name: "Plan", description: "Create a structured plan with steps.", handler_key: "plan" },
    { name: "Plan Enter", description: "Signal the start of a planning phase.", handler_key: "plan_enter" },
    { name: "Plan Exit", description: "Signal the end of a planning phase.", handler_key: "plan_exit" },
    { name: "Question", description: "Ask the user a clarifying question.", handler_key: "question" },
    { name: "Truncate", description: "Truncate content to a max length.", handler_key: "truncate" },
    { name: "Truncation Dir", description: "List directory entries with truncation.", handler_key: "truncation_dir" },
    { name: "Skill Invoke", description: "Invoke another skill by name.", handler_key: "skill" },
  ];
  const tools = [];
  for (const t of toolSpecs) {
    tools.push(await prisma.tool.create({ data: t }));
  }

  for (let i = 0; i < agents.length; i++) {
    await prisma.agent.update({
      where: { id: agents[i].id },
      data: {
        agentSkills: { create: { skill_id: skills[i % skills.length].id } },
        agentTools: { create: { tool_id: tools[i % tools.length].id } },
      },
    });
  }

  const taskTitles = [
    "Set up repo",
    "Design schema",
    "Implement models",
    "Write migrations",
    "Build API routes",
    "Add auth middleware",
    "Create React app",
    "Set up Tailwind",
    "Build Kanban board",
    "Add drag and drop",
    "Integrate WebSockets",
    "Write unit tests",
    "Write integration tests",
    "Set up CI",
    "Deploy to staging",
    "Performance audit",
    "Add logging",
    "Error handling",
    "Write documentation",
    "User feedback loop",
    "Finalize release",
    "Celebrate launch",
  ];
  for (let idx = 0; idx < taskTitles.length; idx++) {
    const col = columns[idx % columns.length];
    const agent = agents[idx % agents.length];
    await prisma.task.create({
      data: {
        project_id: project.id,
        column_id: col.id,
        title: taskTitles[idx],
        description: `Details for ${taskTitles[idx]}`,
        priority: (idx % 4) + 1,
        assigned_agent_id: agent.id,
      },
    });
  }

  const task0 = await prisma.task.findFirst({ where: { project_id: project.id }, orderBy: { id: "asc" } });
  const task1 = await prisma.task.findFirst({ where: { project_id: project.id }, orderBy: { id: "asc" }, skip: 1 });
  const task2 = await prisma.task.findFirst({ where: { project_id: project.id }, orderBy: { id: "asc" }, skip: 2 });
  if (task0) await prisma.taskComment.create({ data: { task_id: task0.id, author: "Alice", body: "Great start!" } });
  if (task1) await prisma.taskComment.create({ data: { task_id: task1.id, author: "Bob", body: "Needs more indexes." } });
  if (task2) await prisma.taskComment.create({ data: { task_id: task2.id, author: "Charlie", body: "Looks good to me." } });

  console.log(`Seeded ${taskTitles.length} tasks, ${agents.length} agents, ${tools.length} tools`);
}

main()
  .catch((e) => {
    console.error(e);
    process.exit(1);
  })
  .finally(async () => {
    await prisma.$disconnect();
  });
